Thanks for the review!

> /Volumes/Data/worker/tvOS-17-Simulator-Build-EWS/build/WebKitBuild/Release-appletvsimulator/SDKAdditions/System/Library/Frameworks/AudioToolbox.framework/Headers/AudioToolbox.h:30:10: fatal error: 'AudioToolbox/AudioServices.h' file not found

The tvOS simulator build failure doesn't seem related.

--- a/LayoutTests/editing/selection/ios/selection-clip-rect-in-overflow-scroller-expected.txt
+++ b/LayoutTests/editing/selection/ios/selection-clip-rect-in-overflow-scroller-expected.txt
@@ -1,0 +1,17 @@
+Verifies that the selection highlight is clipped to overflow scrolling container. Select the first word in the container below and scroll down
+
+On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".
+
+Here’s to the crazy ones. The misfits. The rebels. The troublemakers. The round pegs in the square holes. The ones who see things differently. They’re not fond of rules. And they have no respect for the status quo.
+
+You can quote them, disagree with them, glorify or vilify them. About the only thing you can’t do is ignore them.
+
+Because they change things. They push the human race forward. And while some may see them as the crazy ones, we see genius. Because the people who are crazy enough to think they can change the world, are the ones who do.
+
+PASS getSelection().toString() became 'Here’s'
+PASS selectionRects[0]?.width || 0 is 0
+PASS selectionRects[0]?.height || 0 is 0
+PASS successfullyParsed is true
+
+TEST COMPLETE
+

--- a/LayoutTests/editing/selection/ios/selection-clip-rect-in-overflow-scroller.html
+++ b/LayoutTests/editing/selection/ios/selection-clip-rect-in-overflow-scroller.html
@@ -1,0 +1,70 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ useFlexibleViewport=true AsyncOverflowScrollingEnabled=true ] -->
+<html>
+<head>
+<meta name="viewport" content="width=device-width, initial-scale=1">
+<meta charset="utf-8">
+<script src="../../../resources/ui-helper.js"></script>
+<script src="../../../resources/js-test.js"></script>
+<style>
+body, html {
+    font-size: 16px;
+    font-family: system-ui;
+}
+
+#target {
+    border: 1px solid tomato;
+    padding: 3px;
+    font-size: 24px;
+}
+
+.container {
+    width: 300px;
+    height: 350px;
+    border: solid 1px black;
+    border-radius: 4px;
+    box-sizing: border-box;
+    overflow-y: scroll;
+    line-height: 1.5em;
+    outline: none;
+    padding: 1em;
+}
+</style>
+<script>
+jsTestIsAsync = true;
+
+addEventListener("load", async () => {
+    description("Verifies that the selection highlight is clipped to overflow scrolling container. Select the first word in the container below and scroll down");
+
+    let scroller = document.querySelector(".container");
+    await UIHelper.longPressElement(document.getElementById("target"));
+    await shouldBecomeEqual("getSelection().toString()", "'Here’s'");
+    await UIHelper.waitForSelectionToAppear();
+
+    let {x, y} = UIHelper.midPointOfRect(scroller.getBoundingClientRect());
+    while (scroller.scrollTop < 80) {
+        await UIHelper.sendEventStream(new UIHelper.EventStreamBuilder()
+            .begin(x, y + 60)
+            .move(x, y - 60, 0.25)
+            .end()
+            .takeResult());
+    }
+
+    await UIHelper.waitForZoomingOrScrollingToEnd();
+
+    selectionRects = await UIHelper.getUISelectionViewRects();
+    shouldBe("selectionRects[0]?.width || 0", "0");
+    shouldBe("selectionRects[0]?.height || 0", "0");
+    finishJSTest();
+});
+</script>
+</head>
+<body>
+    <div id="description"></div>
+    <div class="container">
+        <p><span id="target">Here’s</span> to the crazy ones. The misfits. The rebels. The troublemakers. The round pegs in the square holes. The ones who see things differently. They’re not fond of rules. And they have no respect for the status quo.</p>
+        <p>You can quote them, disagree with them, glorify or vilify them. About the only thing you can’t do is ignore them.</p>
+        <p>Because they change things. They push the human race forward. And while some may see them as the crazy ones, we see genius. Because the people who are crazy enough to think they can change the world, are the ones who do.</p>
+    </div>
+    <div id="console"></div>
+</body>
+</html>

--- a/LayoutTests/editing/selection/ios/selection-handle-clamping-in-iframe.html
+++ b/LayoutTests/editing/selection/ios/selection-handle-clamping-in-iframe.html
@@ -52,12 +52,18 @@ async function runTest() {
     }
 
     const [startX, startY] = midPointOfRect(startRect);
-    await touchAndDragFromPointToPoint(startX, startY, startX, startY + 100);
-    await liftUpAtPoint(startX, startY + 100);
+    await UIHelper.sendEventStream(new UIHelper.EventStreamBuilder()
+        .begin(startX, startY)
+        .move(startX, startY + 150, 0.5)
+        .end()
+        .takeResult());
 
     const [endX, endY] = midPointOfRect(endRect);
-    await touchAndDragFromPointToPoint(endX, endY, endX, endY - 100);
-    await liftUpAtPoint(endX, endY - 100);
+    await UIHelper.sendEventStream(new UIHelper.EventStreamBuilder()
+        .begin(endX, endY)
+        .move(endX, endY - 150, 0.5)
+        .end()
+        .takeResult());
 
     result.textContent = target.contentWindow.getSelection().toString();
     testRunner.notifyDone();

--- a/Source/WebCore/dom/Position.h
+++ b/Source/WebCore/dom/Position.h
@@ -250,7 +250,7 @@ Position lastPositionInNode(Node* anchorNode);
 
 bool offsetIsBeforeLastNodeOffset(unsigned offset, Node* anchorNode);
 
-Node* commonInclusiveAncestor(const Position&, const Position&);
+WEBCORE_EXPORT Node* commonInclusiveAncestor(const Position&, const Position&);
 
 WTF::TextStream& operator<<(WTF::TextStream&, const Position&);
 

--- a/Source/WebCore/rendering/RenderObject.h
+++ b/Source/WebCore/rendering/RenderObject.h
@@ -734,7 +734,7 @@ public:
     inline bool preservesNewline() const;
 
     RenderView& view() const { return *document().renderView(); }
-    CheckedRef<RenderView> checkedView() const;
+    WEBCORE_EXPORT CheckedRef<RenderView> checkedView() const;
 
     HostWindow* hostWindow() const;
 
@@ -767,7 +767,7 @@ public:
     // Returns the object containing this one. Can be different from parent for positioned elements.
     // If repaintContainer and repaintContainerSkipped are not null, on return *repaintContainerSkipped
     // is true if the renderer returned is an ancestor of repaintContainer.
-    RenderElement* container() const;
+    WEBCORE_EXPORT RenderElement* container() const;
     RenderElement* container(const RenderLayerModelObject* repaintContainer, bool& repaintContainerSkipped) const;
 
     RenderBoxModelObject* offsetParent() const;

--- a/Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm
+++ b/Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm
@@ -665,13 +665,13 @@ void WebPage::getPlatformEditorStateCommon(const LocalFrame& frame, EditorState&
         }();
     }
 
-    if (RefPtr editableRootOrFormControl = enclosingTextFormControl(selection.start()) ?: selection.rootEditableElement()) {
-#if PLATFORM(IOS_FAMILY)
-        auto& visualData = *result.visualData;
-        visualData.selectionClipRect = rootViewInteractionBounds(*editableRootOrFormControl);
-#endif
+    RefPtr editableRootOrFormControl = enclosingTextFormControl(selection.start()) ?: selection.rootEditableElement();
+    if (editableRootOrFormControl)
         postLayoutData.editableRootIsTransparentOrFullyClipped = result.isContentEditable && isTransparentOrFullyClipped(*editableRootOrFormControl);
-    }
+
+#if PLATFORM(IOS_FAMILY)
+    computeSelectionClipRect(result, selection, editableRootOrFormControl.get());
+#endif
 }
 
 void WebPage::getPDFFirstPageSize(WebCore::FrameIdentifier frameID, CompletionHandler<void(WebCore::FloatSize)>&& completionHandler)

--- a/Source/WebKit/WebProcess/WebPage/WebPage.h
+++ b/Source/WebKit/WebProcess/WebPage/WebPage.h
@@ -1868,6 +1868,7 @@ private:
     void clearSelectionAfterTapIfNeeded();
     void scheduleLayoutViewportHeightExpansionUpdate();
     void scheduleEditorStateUpdateAfterAnimationIfNeeded(const WebCore::Element&);
+    void computeSelectionClipRect(EditorState&, const WebCore::VisibleSelection&, const WebCore::Element* editableRootOrFormControl) const;
 #endif // PLATFORM(IOS_FAMILY)
 
 #if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)

--- a/Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm
+++ b/Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm
@@ -142,6 +142,7 @@
 #import <WebCore/RenderBoxInlines.h>
 #import <WebCore/RenderImage.h>
 #import <WebCore/RenderLayer.h>
+#import <WebCore/RenderObjectInlines.h>
 #import <WebCore/RenderThemeIOS.h>
 #import <WebCore/RenderVideo.h>
 #import <WebCore/RenderView.h>
@@ -5546,6 +5547,70 @@ void WebPage::shouldDismissKeyboardAfterTapAtPoint(FloatPoint point, CompletionH
     bool isReplaced;
     FloatSize targetSize = target->absoluteBoundingRect(&isReplaced).size();
     completion(targetSize.width() >= minimumSizeForDismissal.width() && targetSize.height() >= minimumSizeForDismissal.height());
+}
+
+static CheckedPtr<RenderBox> enclosingScroller(const VisibleSelection& selection)
+{
+    RefPtr ancestor = commonInclusiveAncestor(selection.start(), selection.end());
+    if (!ancestor)
+        return { };
+
+    CheckedPtr selectionRenderer = ancestor->renderer();
+    if (!selectionRenderer)
+        return { };
+
+    auto containingRenderer = [](const RenderObject& renderer) -> CheckedPtr<RenderElement> {
+        if (CheckedPtr container = renderer.container())
+            return container;
+
+        if (RefPtr owner = renderer.protectedDocument()->ownerElement())
+            return owner->renderer();
+
+        return { };
+    };
+
+    CheckedPtr candidate = dynamicDowncast<RenderElement>(selectionRenderer.get()) ?: selectionRenderer->container();
+    for (; candidate; candidate = containingRenderer(*candidate)) {
+        if (CheckedPtr renderBox = dynamicDowncast<RenderBox>(*candidate); renderBox && renderBox->canBeScrolledAndHasScrollableArea())
+            return renderBox;
+    }
+
+    return { };
+}
+
+void WebPage::computeSelectionClipRect(EditorState& state, const VisibleSelection& selection, const Element* editableRootOrFormControl) const
+{
+    if (editableRootOrFormControl)
+        state.visualData->selectionClipRect = rootViewInteractionBounds(Ref { *editableRootOrFormControl });
+
+    if (selection.isNoneOrOrphaned())
+        return;
+
+    if (!state.isContentEditable && selection.isCaret())
+        return;
+
+    CheckedPtr scroller = enclosingScroller(selection);
+    if (!scroller)
+        return;
+
+    Ref view = scroller->checkedView()->frameView();
+    IntRect scrollerClipRectInContent;
+    if (CheckedPtr renderView = dynamicDowncast<RenderView>(*scroller)) {
+        if (renderView->protectedDocument()->isTopDocument())
+            return;
+
+        scrollerClipRectInContent = view->visibleContentRect();
+    } else
+        scrollerClipRectInContent = scroller->absoluteBoundingBoxRect();
+
+    if (scrollerClipRectInContent.isEmpty())
+        return;
+
+    auto scrollerClipRectInRootView = view->contentsToRootView(scrollerClipRectInContent);
+    if (state.visualData->selectionClipRect.isEmpty())
+        state.visualData->selectionClipRect = scrollerClipRectInRootView;
+    else
+        state.visualData->selectionClipRect.intersect(scrollerClipRectInRootView);
 }
 
 } // namespace WebKit

--- a/Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm
+++ b/Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm
@@ -26,6 +26,7 @@
 #import "config.h"
 #import "TestRunnerWKWebView.h"
 
+#import "PlatformViewHelpers.h"
 #import "TestController.h"
 #import "WebKitTestRunnerDraggingInfo.h"
 #import <WebKit/WKUIDelegatePrivate.h>
@@ -384,14 +385,11 @@ IGNORE_WARNINGS_END
     }
 }
 
-- (BOOL)isZoomingOrScrolling
+static bool isZoomingOrScrolling(UIScrollView *scroller)
 {
-    auto scroller = self.scrollView;
     if (scroller.isZooming || scroller.isAnimatingZoom || scroller.isAnimatingScroll
-        || scroller.isVerticalBouncing || scroller.isHorizontalBouncing || scroller.isZoomBouncing)
-        return YES;
-
-    if (self._keyboardScrollingAnimationRunning)
+        || scroller.isVerticalBouncing || scroller.isHorizontalBouncing || scroller.isZoomBouncing
+        || scroller.decelerating || scroller.dragging || scroller.scrollAnimating || scroller.tracking)
         return YES;
 
     static NeverDestroyed<RetainPtr<NSSet>> animationKeyNames = [NSSet setWithArray:@[
@@ -403,6 +401,19 @@ IGNORE_WARNINGS_END
 
     for (NSString *key in scroller.layer.animationKeys) {
         if ([animationKeyNames.get() containsObject:key])
+            return YES;
+    }
+
+    return NO;
+}
+
+- (BOOL)isZoomingOrScrolling
+{
+    if (self._keyboardScrollingAnimationRunning)
+        return YES;
+
+    for (UIScrollView *scroller in findAllViewsInHierarchyOfType(self, UIScrollView.class)) {
+        if (isZoomingOrScrolling(scroller))
             return YES;
     }
 

